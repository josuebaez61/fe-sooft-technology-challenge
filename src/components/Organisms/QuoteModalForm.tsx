import Modal from "./Modal";
import Button from "../Atoms/Button";
import { useForm } from "react-hook-form";
import type {
  CreateQuotePayload,
  QuoteForm as QuoteFormModel,
} from "../../models";
import QuoteForm from "../Molecules/QuoteForm";
import { useEffect } from "react";

interface QuoteModalFormProps {
  visible?: boolean;
  isCreating?: boolean;
  onCreate: (formValue: CreateQuotePayload) => void;
  onClose?: () => void;
  createError?: string | null;
}

export default function QuoteModalForm({
  isCreating,
  visible,
  onCreate,
  onClose,
  createError,
}: QuoteModalFormProps) {
  const { control, handleSubmit, reset } = useForm<QuoteFormModel>();

  useEffect(() => {
    if (!visible) {
      reset();
    }
  }, [visible, reset]);

  useEffect(() => {
    if (createError) {
      alert(createError);
    }
  }, [createError]);

  const onSubmit = async (formValue: QuoteFormModel) => {
    if (!formValue.author) {
      alert("El autor es requerido");
      return;
    }
    if (!formValue.quote) {
      alert("La cita es requerida");
      return;
    }
    const payload: CreateQuotePayload = {
      author: formValue.author,
      quote: formValue.quote,
    };
    onCreate(payload);
  };

  const renderModalFooter = () => {
    return (
      <div className="flex justify-end gap-2">
        <Button isLoading={isCreating} onClick={handleSubmit(onSubmit)}>
          Guardar
        </Button>
        <Button disabled={isCreating} onClick={onClose}>
          Cancelar
        </Button>
      </div>
    );
  };
  return (
    <Modal
      title="Nueva frase"
      visible={visible}
      footer={renderModalFooter}
      onClose={onClose}
    >
      <QuoteForm control={control} />
    </Modal>
  );
}
