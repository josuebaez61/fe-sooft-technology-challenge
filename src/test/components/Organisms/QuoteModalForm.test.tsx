import { render } from "@testing-library/react";
import QuoteModalForm from "../../../components/Organisms/QuoteModalForm";

describe("QuoteModalForm", () => {
  const onCreate = jest.fn();
  const onClose = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("should match snapshot", () => {
    const { container } = render(
      <QuoteModalForm
        visible={true}
        onCreate={onCreate}
        onClose={onClose}
        isCreating={false}
        createError={null}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("shows alert if createError changes", () => {
    const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => {});
    const { rerender } = render(
      <QuoteModalForm
        visible={true}
        onCreate={onCreate}
        onClose={onClose}
        isCreating={false}
        createError={null}
      />
    );
    rerender(
      <QuoteModalForm
        visible={true}
        onCreate={onCreate}
        onClose={onClose}
        isCreating={false}
        createError="Test error"
      />
    );
    expect(alertSpy).toHaveBeenCalledWith("Test error");
    alertSpy.mockRestore();
  });
});
