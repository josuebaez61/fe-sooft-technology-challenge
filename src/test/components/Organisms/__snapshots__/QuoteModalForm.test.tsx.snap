// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteModalForm should match snapshot 1`] = `
<div>
  <div
    class="animate__animated animate__fadeIn animate__faster"
  >
    <div
      class="fixed inset-0 bg-backdrop bg-opacity-50"
    >
      <div
        class="flex items-center justify-center h-screen"
      >
        <div
          class="relative"
        >
          <div
            class="card border border-golden p-5 flex flex-col w-full min-w-[280px] sm:min-w-[340px] md:min-w-[400px] max-w-lg sm:max-w-xl md:max-w-2xl lg:max-w-3xl xl:max-w-4xl"
          >
            <div
              class="card-header mb-4 relative"
            >
              <div
                class="flex justify-between"
              >
                <h2
                  class="text-2xl"
                >
                  <span
                    class="text-golden"
                  >
                    Nueva frase
                  </span>
                </h2>
                <button
                  aria-label="Cerrar"
                  class="text-gray-400 hover:text-gray-600 text-xl cursor-pointer"
                >
                  Ã—
                </button>
              </div>
            </div>
            <div
              class="card-body"
            >
              <div
                class="flex gap-2 flex-col py-2 w-full"
              >
                <label
                  class="text-sm font-semibold"
                >
                  <span
                    class="text-golden"
                  >
                    Frase
                  </span>
                </label>
                <textarea
                  class="w-full appearance-none outline-none border-2 border-golden bg-transparent text-white p-2 rounded-none transition-colors"
                  name="quote"
                />
              </div>
              <div
                class="flex gap-2 flex-col py-2 w-full"
              >
                <label
                  class="text-sm font-semibold"
                >
                  <span
                    class="text-golden"
                  >
                    Autor
                  </span>
                </label>
                <input
                  class="w-full appearance-none outline-none border-2 border-golden bg-transparent text-white p-2 rounded-none transition-colors"
                  name="author"
                  value=""
                />
              </div>
            </div>
            <div
              class="card-footer mt-4"
            >
              <div
                class="flex justify-end gap-2"
              >
                <button
                  class="button border border-golden py-2 px-3 flex items-center justify-center gap-2 cursor-pointer"
                >
                  <span
                    class="text-golden inline-block"
                  >
                    Guardar
                  </span>
                </button>
                <button
                  class="button border border-golden py-2 px-3 flex items-center justify-center gap-2 cursor-pointer"
                >
                  <span
                    class="text-golden inline-block"
                  >
                    Cancelar
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
